pkg/services/ngalert/state/historian/fanout.go

```diff
@@ -0,0 +1,52 @@
+package historian
+
+import (
+	"context"
+	"errors"
+
+	"github.com/grafana/grafana-plugin-sdk-go/data"
+	ngmodels "github.com/grafana/grafana/pkg/services/ngalert/models"
+	"github.com/grafana/grafana/pkg/services/ngalert/state"
+	history_model "github.com/grafana/grafana/pkg/services/ngalert/state/historian/model"
+)
+
+type Backend interface {
+	Record(ctx context.Context, rule history_model.RuleMeta, states []state.StateTransition) <-chan error
+	Query(ctx context.Context, query ngmodels.HistoryQuery) (*data.Frame, error)
+}
+
+// FanoutBackend is a state.Historian that records history to multiple other backends at once.
+// Only one backend is used for reads. The backend selected for read traffic is called the primary and all others are called secondaries.
+type FanoutBackend struct {
+	targets []Backend
+}
+
+func NewFanoutBackend(primary Backend, secondaries ...Backend) *FanoutBackend {
+	return &FanoutBackend{
+		targets: append([]Backend{primary}, secondaries...),
+	}
+}
+
+func (h *FanoutBackend) Record(ctx context.Context, rule history_model.RuleMeta, states []state.StateTransition) <-chan error {
+	writes := make([]<-chan error, 0, len(h.targets))
+	for _, b := range h.targets {
+		writes = append(writes, b.Record(ctx, rule, states))
+	}
+	errCh := make(chan error, 1)
+	go func() {
+		defer close(errCh)
+		errs := make([]error, 0)
```

Comment:
It took me a while to realise that no matter how we consume the channels, we'll always need to wait on the slowest one to report back - I'd add some comments around this area.